### PR TITLE
EWB-1744: Raise error when attempting to configure auth on EasClient for HTTP server

### PR DESCRIPTION
**Pull request by Lax125**
Signed-off-by: Marcus Koh <marcus.koh@zepben.com>

### CODE REVIEW

**Comment by ryanjcoleman** [comment]
Task linked: [EWB-1744 Raise error in EasClient if initialised with auth but non https protocol](https://app.clickup.com/t/6929263/EWB-1744)

File: src/zepben/eas/client/eas_client.py [hunk 1]
```diff
@@ -30,45 +30,73 @@ def __init__(
...
         ca_filename: Optional[str] = None,
         session: ClientSession = None,
         json_serialiser = None
     ):
         """
-        :param host: The host string of the Evolve App Server, including the protocol, e.g."https://evolve.local"
+        Construct a client for the Evolve App Server. If the server is HTTPS, authentication may be configured.
+        Authentication may be configured in one of two ways:
+            - Specifying a ZepbenTokenFetcher directly via the token_fetcher parameter
+            - Specifying the client ID of the Auth0 via the client_id parameter, plus one of the following:
```
Comment (kgreav) [inline src/zepben/eas/client/eas_client.py:48]: nit: `of the Auth0 application`

File: src/zepben/eas/client/eas_client.py [hunk 2]
```diff
@@ -30,45 +30,73 @@ def __init__(
...
+                - The client secret via the client_secret parameter (M2M authentication)
+                - A username and password pair via the username and password parameters (account authentication)
+              If this method id used, the auth configuration will be fetched from the Evolve App Server at the path
```
Comment (kgreav) [inline src/zepben/eas/client/eas_client.py:51]: nit: `if this method is used`

File: src/zepben/eas/client/eas_client.py [hunk 3]
```diff
@@ -30,45 +30,73 @@ def __init__(
...
+              "/api/config/auth".
+
+        Address parameters:
+        :param host: The domain of the Evolve App Server, including the protocol, e.g. "evolve.local"
```
Comment (kgreav) [inline src/zepben/eas/client/eas_client.py:55]: this doesn't include the protocol.

File: src/zepben/eas/client/eas_client.py (hunk 1 above)
Comment (kgreav) [reply src/zepben/eas/client/eas_client.py:48]: Can you also re-order these? We prefer people use the client id + username/password or secret over the zepben token fetcher, so better to make the first thing in the list the one most relevant to users.

File: src/zepben/eas/client/eas_client.py [hunk 4]
```diff
@@ -30,45 +30,73 @@ def __init__(
...
-        :param verify_certificate: Set this to False to disable certificate verification
-        :param ca_filename: Path to CA file to use for verification.
-        :param session: aiohttp CLientSession to use, if not provided a new session will be created for you. You should typically only use one aiohttp session
-        per application.
-        :param json_serialiser: JSON serialiser to use for requests e.g ujson.dumps. Defaults to json.dumps
+        :param protocol: The protocol of the Evolve App Server. Should be either "http" or "https". Must be "https" if
+                         auth is configured. (Defaults to "https")
+
+        Authentication parameters:
+        :param token_fetcher: A ZepbenTokenFetcher used to fetch auth tokens for access to the Evolve App Server.
```
Comment (kgreav) [inline src/zepben/eas/client/eas_client.py:61]: Also here move this to below all the username/password options as its not preferred

File: tox.ini [hunk 5]
```diff
@@ -9,6 +9,8 @@ pip_pre = true
 deps =
     .[test]
 commands =
+    pytest --cov=zepben.eas --cov-report=xml --cov-branch
+    bash -ec 'if [[ -n $COVERALLS_REPO_TOKEN ]]; then coveralls; fi'
```
Comment (kgreav) [inline tox.ini:13]: we don't need coveralls anymore, we're using codecov, which you should be able to change to look like the [evolve-sdk-python tox.ini](https://github.com/zepben/evolve-sdk-python/blob/main/tox.ini)

**Review by kgreav — CHANGES_REQUESTED** [review]

**Comment by kgreav** [comment]
Looks good, I just added some extra checks to force the user into picking one set of creds so that we don't have ambiguous behaviour if you mistakenly send both (e.g token fetcher and username+password).

**Review by kgreav — APPROVED** [review]

**Pull request merged**